### PR TITLE
add attributes for optional datadog event

### PR DESCRIPTION
**Pull request by slantin**
**What are you trying to accomplish with this PR?**
Add a way to optionally send an event to Datadog via statsd on update. This will open up the doors to sending an event for nginx binary deploy as per Shopify/traffic#962 in the `nginx` cookbook.

**What could go wrong? (if anything)**
Event could not get sent

**Rollback steps**
- [x] It is safe to simply rollback this change.
- [x] Users will not be locked out of the nodes (please indicate if users will be locked out)

**Checklist**
- [ ] I have tested my changes locally in test kitchen
- [x] I will make sure my changes pass CI before merging
- [ ] I have :tophat:'d these changes

Reviewers @Shopify/traffic @marc-barry @klautcomputing 
Issue Shopify/traffic#962

### CODE REVIEW

File: providers/default.rb [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
 
-  dir = new_resource.base_name.split('/')[0..-2].join('/')
   directory dir
 
   file new_resource.base_name do
     content new_resource.version
     notifies :run, 'execute[apt-get-update]', :immediately
   end
+
+  if push_event
```
Comment (klautcomputing) [inline providers/default.rb]: `if push_event && event_name && event_text`

File: providers/default.rb [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
   execute 'apt-get-update' do
     command 'apt-get -q update'
     action :nothing
   end
 
-  dir = new_resource.base_name.split('/')[0..-2].join('/')
   directory dir
 
   file new_resource.base_name do
     content new_resource.version
```
Comment (klautcomputing) [inline providers/default.rb]: not really your code, but I would move `version` up there with all the other initialization

File: resources/default.rb [hunk 3]
```diff
@@ -4,3 +4,7 @@
 
 attribute :base_name, name_attribute: true, kind_of: String, required: true
 attribute :version, kind_of: String, required: true
+attribute :push_to_datadog, kind_of: [TrueClass, FalseClass], required: false
+attribute :datadog_event_name, kind_of: String, required: false
```
Comment (klautcomputing) [inline resources/default.rb]: add a `default: nil` to all your strings to get the code up there working correctly

File: dev.yml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-name: shopify_apt_get_on_change
+name: shopify-apt-get-on-change
```
Comment (klautcomputing) [inline dev.yml]: why has that a changed?

File: providers/default.rb [hunk 5]
```diff
@@ -1,16 +1,30 @@
...
     notifies :run, 'execute[apt-get-update]', :immediately
   end
+
+  if push_event
+    unless event_name.nil? || event_text.nil?
+      require 'datadog/statsd'
+      statsd = Datadog::Statsd.new('localhost', 8125)
+      statsd.event(datadog_event_name, datadog_event_text,
```
Comment (klautcomputing) [inline providers/default.rb]: we do max 120 char per line, do you do that in your editor as well?

File: dev.yml (hunk 4 above)
Comment (slantin) [inline dev.yml]: running `dev up` changed it automatically

```→ dev up
🦄  dev.yml just changed. you may want to run dev up.
🦄 💥  The dev.yml name value must be made up of only a-z, 0-9, and -.
🦄 💥  The corrected value was written into your dev.yml, please commit it.
```

Comment (klautcomputing) [inline dev.yml]: cool, the more you know 👍

File: providers/default.rb (hunk 5 above)
Comment (slantin) [inline providers/default.rb]: `bundle exec rake style` seems to tell me 80 characters is the maximum when I have it all on one line

File: providers/default.rb [hunk 6]
```diff
@@ -1,16 +1,30 @@
 use_inline_resources
 
 action :update do
+  dir        = new_resource.base_name.split('/')[0..-2].join('/')
+  push_event = new_resource.push_to_datadog || false
```
Comment (ElvinEfendi) [inline providers/default.rb]: The default value should be defined at the resource definition.

File: providers/default.rb [hunk 7]
```diff
@@ -1,16 +1,30 @@
...
+  event_name = new_resource.datadog_event_name
+  event_text = new_resource.datadog_event_text
+  event_key  = new_resource.datadog_event_key
+
```
Comment (ElvinEfendi) [inline providers/default.rb]: Why do you introduce all these variables instead of using `new_resource.<attribute_name>`?

Comment (klautcomputing) [inline providers/default.rb]: We seem to do it this way in a number of cookbooks, so I assumed that it's the way to go. And I showed @slantin some code where they did it this way.

Comment (slantin) [inline providers/default.rb]: felix mentioned that this was better style

Comment (slantin) [inline providers/default.rb]: ^

Comment (ElvinEfendi) [inline providers/default.rb]: @klautcomputing I could not really see that pattern in our custom resources and also among the open source LWRPs we use at cookbooks-upstream. In this situation I just do not see any value of doing this.

File: providers/default.rb [hunk 8]
```diff
@@ -1,16 +1,28 @@
 use_inline_resources
 
 action :update do
+  dir        = new_resource.base_name.split('/')[0..-2].join('/')
+  version    = new_resource.version
+  push_event = new_resource.push_to_datadog
+  event_name = new_resource.datadog_event_name
+  event_text = new_resource.datadog_event_text
+  event_key  = new_resource.datadog_event_key
+
```
Comment (ElvinEfendi) [inline providers/default.rb]: Have you thought of providing something like the following interface instead?
```
shopify_apt_get_on_change '/etc/nginx/version' do
  version node['nginx']['version']
  datadog_event: {
    name: "",
    text: "",
    key: ""
  }
end
```
I think this is quite common Chef pattern.

File: resources/default.rb [hunk 9]
```diff
@@ -4,3 +4,7 @@
 
 attribute :base_name, name_attribute: true, kind_of: String, required: true
 attribute :version, kind_of: String, required: true
+attribute :push_to_datadog, kind_of: [TrueClass, FalseClass], required: false, default: false
+attribute :datadog_event_name, kind_of: String, required: false, default: nil
+attribute :datadog_event_text, kind_of: String, required: false, default: nil
+attribute :datadog_event_key, kind_of: String, required: false, default: nil
```
Comment (ElvinEfendi) [inline resources/default.rb]: Have you thought of providing something like the following interface instead?
```
shopify_apt_get_on_change '/etc/nginx/version' do
  version node['nginx']['version']
  datadog_event: {
    name: "",
    text: "",
    key: ""
  }
end
```
I think this is quite common Chef pattern.

Comment (slantin) [inline resources/default.rb]: that's definitely much neater, I think I will do that

File: libraries/datadog_push.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+require 'datadog/statsd'
+
+def push_to_datadog(event)
+  if event && event[:name] && event [:text]
```
Comment (ElvinEfendi) [inline libraries/datadog_push.rb]: extra space?

File: spec/default_spec.rb [hunk 11]
```diff
@@ -7,7 +7,10 @@
                         .converge(described_recipe)
   end
 
+  before { allow_any_instance_of(Chef::Provider).to receive(:push_to_datadog) }
```
Comment (ElvinEfendi) [inline spec/default_spec.rb]: Why is this necessary?

Comment (slantin) [inline spec/default_spec.rb]: this was to stub out the actual statsd code

File: libraries/datadog_push.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+require 'datadog/statsd'
+
+def push_to_datadog(event)
+  if event && event[:name] && event[:text]
```
Comment (ElvinEfendi) [inline libraries/datadog_push.rb]: Would be clearer if you move this to the provider itself and use chef's built-in `if{}` construct instead of this `if`.

File: spec/default_spec.rb [hunk 13]
```diff
@@ -7,7 +7,10 @@
...
+
   it 'should update shopify_apt_get_on_change' do
+    puts chef_run.node.normal
```
Comment (ElvinEfendi) [inline spec/default_spec.rb]: debug statement should be removed before merging
